File: functions/src/researcher-dashboard/portal-token.test.ts
import jwt from "jsonwebtoken"
import { generateKeyPairSync, createHmac } from "crypto"
import { execFileSync } from "child_process"
import { parsePortalKeys, verifyPortalToken, PortalTokenError, PortalKeys } from "./portal-token"

const AUDIENCE = "report-service-functions"

const makeKey = (kid: string, iss: string) => {
  const { publicKey, privateKey } = generateKeyPairSync("rsa", { modulusLength: 2048 })
  return {
    kid,
    iss,
    publicPem: publicKey.export({ type: "spki", format: "pem" }).toString(),
    privatePem: privateKey.export({ type: "pkcs8", format: "pem" }).toString()
  }
}

const staging = makeKey("staging-test", "https://learn.portal.staging.concord.org/")
const production = makeKey("production-test", "https://learn.concord.org/")

const keysJson = JSON.stringify([staging, production].map(k => ({ kid: k.kid, iss: k.iss, pem: k.publicPem })))
const keys: PortalKeys = parsePortalKeys(keysJson)

const now = () => Math.floor(Date.now() / 1000)
const claims = (overrides: Record<string, unknown> = {}) =>
  ({ iss: staging.iss, aud: AUDIENCE, uid: 42, iat: now(), exp: now() + 120, ...overrides })

// kid null signs with no kid in the header
const sign = (payload: object, key = staging, kid: string | null = key.kid) =>
  jwt.sign(payload, key.privatePem, { algorithm: "RS256", ...(kid ? { keyid: kid } : {}) })

const b64url = (value: object) => Buffer.from(JSON.stringify(value)).toString("base64url")

const hmacWithPublicPem = (payload: object, headerAlg: string) => {
  const input = `${b64url({ alg: headerAlg, typ: "JWT", kid: staging.kid })}.${b64url(payload)}`
  const signature = createHmac("sha256", staging.publicPem).update(input).digest("base64url")
  return `${input}.${signature}`
}

const expectRefused = (token: string, message?: RegExp) => {
  expect(() => verifyPortalToken(token, AUDIENCE, keys)).toThrow(PortalTokenError)
  if (message) expect(() => verifyPortalToken(token, AUDIENCE, keys)).toThrow(message)
}

describe("verifyPortalToken", () => {
  it("accepts a token signed by the key its kid names, for that key's issuer", () => {
    expect(verifyPortalToken(sign(claims()), AUDIENCE, keys)).toMatchObject({ uid: 42, iss: staging.iss })
  })

  it("refuses the staging key claiming the production issuer", () => {
    expectRefused(sign(claims({ iss: production.iss })), /issuer invalid/)
  })

  it("refuses the production key under the staging kid", () => {
    expectRefused(sign(claims(), production, staging.kid), /invalid signature/)
  })

  it("refuses an unknown kid rather than falling back to a configured key", () => {
    expectRefused(sign(claims(), staging, "retired-2025"), /unknown or missing kid/)
  })

  it("refuses a token with no kid", () => {
    expectRefused(sign(claims(), staging, null), /unknown or missing kid/)
  })

  it("refuses HS256 signed with the configured public key as the HMAC secret", () => {
    expectRefused(hmacWithPublicPem(claims(), "HS256"), /invalid algorithm/)
  })

  it("refuses an HMAC signature under a header that claims RS256", () => {
    expectRefused(hmacWithPublicPem(claims(), "RS256"), /invalid signature/)
  })

  it("refuses alg none", () => {
    expectRefused(`${b64url({ alg: "none", typ: "JWT", kid: staging.kid })}.${b64url(claims())}.`)
  })

  it("refuses another audience", () => {
    expectRefused(sign(claims({ aud: "report-server" })), /audience invalid/)
  })

  it("refuses a token with no aud", () => {
    const { aud, ...rest } = claims()
    expectRefused(sign(rest))
  })

  it("refuses an aud array even when it contains the expected audience", () => {
    expectRefused(sign(claims({ aud: [AUDIENCE] })), /aud must be a single string/)
  })

  it("refuses an expired token", () => {
    expectRefused(sign(claims({ exp: now() - 1 })), /expired/)
  })

  it("refuses a token with no exp", () => {
    const { exp, ...rest } = claims()
    expectRefused(sign(rest), /exp is required/)
  })

  it("refuses a token with no numeric uid", () => {
    expectRefused(sign(claims({ uid: "42" })), /uid is required/)
  })

  it("refuses garbage", () => {
    expectRefused("not.a.jwt")
    expectRefused("")
    expectRefused(`${b64url({ alg: "RS256", typ: "JWT", kid: staging.kid })}.${Buffer.from("not json").toString("base64url")}.sig`)
  })
})

describe("parsePortalKeys", () => {
  it("reads the value firebase-tools produces from a single-quoted .env line", () => {
    // firebase-tools needs a newer module resolver than this Jest, so its parser runs in node
    const script = `const { envs, errors } = require("firebase-tools/lib/functions/env").parse(process.argv[1]);
      process.stdout.write(JSON.stringify({ value: envs.PORTAL_PUBLIC_KEYS, errors }))`
    const parsed = JSON.parse(execFileSync(process.execPath, ["-e", script, `PORTAL_PUBLIC_KEYS='${keysJson}'\n`]).toString())

    expect(parsed.errors).toEqual([])
    expect(Array.from(parsePortalKeys(parsed.value).keys())).toEqual([staging.kid, production.kid])
  })

  it("treats an empty value as no keys", () => {
    expect(parsePortalKeys("").size).toBe(0)
  })

  it("throws on malformed JSON, a non-array, a missing or empty field, a bad PEM or a repeated kid", () => {
    const entry = { kid: staging.kid, iss: staging.iss, pem: staging.publicPem }
    expect(() => parsePortalKeys("{not json")).toThrow()
    expect(() => parsePortalKeys(JSON.stringify(entry))).toThrow(/JSON array/)
    expect(() => parsePortalKeys(JSON.stringify([{ kid: "k", iss: "i" }]))).toThrow(/kid, iss and pem/)
    expect(() => parsePortalKeys(JSON.stringify([{ ...entry, iss: "" }]))).toThrow(/kid, iss and pem/)
    expect(() => parsePortalKeys(JSON.stringify([{ ...entry, pem: "not a pem" }]))).toThrow()
    expect(() => parsePortalKeys(JSON.stringify([entry, entry]))).toThrow(/more than once/)
  })
})

File: server/test/report_server_web/api/portal_token_test.exs
defmodule ReportServerWeb.Api.PortalTokenTest do
  use ReportServerWeb.ConnCase, async: false

  import ReportServerWeb.PortalTokenFixture

  alias ReportServerWeb.Api.{PortalToken, PortalTokenPlug}

  @audience "report-server"

  describe "verify/2" do
    test "accepts a token signed by the key its kid names, for that key's issuer" do
      token = sign(:staging, claims(:staging, @audience))

      assert {:ok, %{"uid" => 42, "aud" => @audience}} = PortalToken.verify(token, @audience)
    end

    test "refuses the staging key claiming the production issuer" do
      token = sign(:staging, claims(:staging, @audience, %{"iss" => iss(:production)}))

      assert PortalToken.verify(token, @audience) == {:error, :wrong_issuer}
    end

    test "refuses the production key under the staging kid" do
      token = sign(:production, claims(:staging, @audience), kid: kid(:staging))

      assert PortalToken.verify(token, @audience) == {:error, :signature_error}
    end

    test "refuses an unknown kid rather than falling back to a configured key" do
      token = sign(:staging, claims(:staging, @audience), kid: "retired-2025")

      assert PortalToken.verify(token, @audience) == {:error, :unknown_kid}
    end

    test "refuses a token with no kid" do
      token = sign(:staging, claims(:staging, @audience)) |> drop_kid()

      assert PortalToken.verify(token, @audience) == {:error, :unsupported_header}
    end

    test "refuses HS256 signed with the configured public key as the HMAC secret" do
      token = sign_hs256_with_public_pem(:staging, claims(:staging, @audience))

      assert PortalToken.verify(token, @audience) == {:error, :unsupported_header}
    end

    test "refuses an HMAC signature under a header that claims RS256" do
      token = sign_hs256_with_public_pem(:staging, claims(:staging, @audience), header_alg: "RS256")

      assert PortalToken.verify(token, @audience) == {:error, :signature_error}
    end

    test "refuses alg none" do
      token = unsigned(:staging, claims(:staging, @audience))

      assert PortalToken.verify(token, @audience) == {:error, :unsupported_header}
    end

    test "refuses another audience" do
      token = sign(:staging, claims(:staging, "researcher-dashboard"))

      assert PortalToken.verify(token, @audience) == {:error, :wrong_audience}
    end

    test "refuses a token with no aud" do
      token = sign(:staging, claims(:staging, @audience), without: ["aud"])

      assert PortalToken.verify(token, @audience) == {:error, :wrong_audience}
    end

    test "refuses an aud list even when it contains the expected audience" do
      token = sign(:staging, claims(:staging, @audience, %{"aud" => [@audience]}))

      assert PortalToken.verify(token, @audience) == {:error, :wrong_audience}
    end

    test "refuses an expired token" do
      token = sign(:staging, claims(:staging, @audience, %{"exp" => System.system_time(:second) - 1}))

      assert PortalToken.verify(token, @audience) == {:error, :expired}
    end

    test "refuses a token with no exp" do
      token = sign(:staging, claims(:staging, @audience), without: ["exp"])

      assert PortalToken.verify(token, @audience) == {:error, :no_expiry}
    end

    test "refuses garbage" do
      assert {:error, _} = PortalToken.verify("not.a.jwt", @audience)
      assert {:error, _} = PortalToken.verify("", @audience)
    end

    @tag :capture_log
    test "trusts no key when PORTAL_PUBLIC_KEYS is malformed" do
      token = sign(:staging, claims(:staging, @audience))

      with_portal_keys("{not json", fn ->
        assert PortalToken.verify(token, @audience) == {:error, :unknown_kid}
      end)
    end
  end

  describe "PORTAL_PUBLIC_KEYS" do
    @describetag :capture_log

    test "ignores an entry whose PEM is unreadable, so its kid is unknown" do
      token = sign(:staging, claims(:staging, @audience))
      keys = Jason.encode!([%{kid: kid(:staging), iss: iss(:staging), pem: "not a pem"}])

      with_portal_keys(keys, fn ->
        assert PortalToken.verify(token, @audience) == {:error, :unknown_kid}
      end)
    end

    test "trusts neither entry when a kid is listed twice, even if the repeat is malformed" do
      token = sign(:staging, claims(:staging, @audience))

      keys =
        Jason.encode!([
          %{kid: kid(:staging), iss: iss(:staging), pem: public_pem(:staging)},
          %{kid: kid(:staging), iss: iss(:production), pem: "not a pem"}
        ])

      with_portal_keys(keys, fn ->
        assert PortalToken.verify(token, @audience) == {:error, :unknown_kid}
      end)
    end

    test "trusts neither entry when a kid is listed twice" do
      token = sign(:staging, claims(:staging, @audience))
      pem = public_pem(:staging)

      keys =
        Jason.encode!([
          %{kid: kid(:staging), iss: iss(:staging), pem: pem},
          %{kid: kid(:staging), iss: iss(:production), pem: pem}
        ])

      with_portal_keys(keys, fn ->
        assert PortalToken.verify(token, @audience) == {:error, :unknown_kid}
      end)
    end
  end

  describe "PortalTokenPlug" do
    test "assigns the verified claims for its audience", %{conn: conn} do
      token = sign(:staging, claims(:staging, @audience))

      conn = call_plug(conn, token, @audience)

      refute conn.halted
      assert %{"uid" => 42} = conn.assigns.portal_claims
    end

    test "answers NOT_AUTHENTICATED for a token of another audience", %{conn: conn} do
      token = sign(:staging, claims(:staging, "researcher-dashboard"))

      conn = call_plug(conn, token, @audience)

      assert conn.halted
      assert json_response(conn, 401)["error"] == "NOT_AUTHENTICATED"
      refute Map.has_key?(conn.assigns, :portal_claims)
    end

    test "answers NOT_AUTHENTICATED with no bearer", %{conn: conn} do
      conn = PortalTokenPlug.call(conn, PortalTokenPlug.init(audience: @audience))

      assert conn.halted
      assert json_response(conn, 401)["error"] == "NOT_AUTHENTICATED"
    end
  end

  defp call_plug(conn, token, audience) do
    conn
    |> put_req_header("authorization", "Bearer #{token}")
    |> PortalTokenPlug.call(PortalTokenPlug.init(audience: audience))
  end

  defp drop_kid(token) do
    [header, payload, signature] = String.split(token, ".")
    decoded = header |> Base.url_decode64!(padding: false) |> Jason.decode!() |> Map.delete("kid")
    "#{Base.url_encode64(Jason.encode!(decoded), padding: false)}.#{payload}.#{signature}"
  end

  defp with_portal_keys(value, fun) do
    previous = Application.get_env(:report_server, :portal_public_keys)
    Application.put_env(:report_server, :portal_public_keys, value)

    try do
      fun.()
    after
      Application.put_env(:report_server, :portal_public_keys, previous)
    end
  end
end
